### PR TITLE
Load main loop plug-ins automatically

### LINKED ISSUE #3558: Load plugins without having to re-install Cylc Flow
**Describe exactly what you would like to see in an upcoming release**

Given that we now have plug-ins in the main loop after #3492 (thanks to @oliver-sanders), we now should be able to work on the plug-in loading.

At the moment it's loading the entrypoints from the `cylc-flow` module, with a specific name. So after creating a new plug-in, users would need to check out the specific version of Cylc Flow they have, update `setup.cfg`, and then re-install it.

Instead, we should be able to navigate the python path and load all the entrypoints that match a specific name. The name would need to contain a namespace too. 

**Additional context**

What we have:

https://github.com/cylc/cylc-flow/blob/506fb35f3a2b26e746c29ab06e79d800a7005492/cylc/flow/main_loop/__init__.py#L312-L321

What we could try:

https://packaging.python.org/guides/creating-and-discovering-plugins/
**Pull requests welcome!**

### PR DESCRIPTION
**Pull request by kinow**
These changes close #3558 

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Does not need tests (why? unfortunately not easy to test this without manually creating a plugin and installing - not impossible though, just hard).
- [x] No change log entry required (why? e.g. invisible to users).
- [x] No documentation update required.

### CODE REVIEW

**Comment by kinow** [comment]
From the Python docs in the linked issue, we have three well-known options for loading plug-ins in Python (though there are modules for that that provide extra options, but not stdlib):

- use namespaces
  * but then we would demand that each user followed that PEP-420, and using namespaces is harder can lead to all sort of issues depending on how you package your code
- use a naming convention for modules
  * I think flask does something like this (though they have some extra more elaborate customization for plug-ins), so users would have to create the modules with names like cylc_etc_my_plugin. But that prevents - as far as I can tell - users from creating plugins in their existing python code, especially if they use namespaces
- use setuptools and package metadata
  * that's what we are using at the moment

We have the third option, which I think is the best one. So this pull request does not change that.

At the moment we load every `main_loop` entry point (i.e. an entry point named `main_loop` in `setup.py` or `setup.cfg`). The only issue is that we load this entry point only from the `cylc-flow` module.

This pull request changes that, and loads the `main_loop` from every module. The only disadvantage I can think of, is that it could be slow (`pkg_resources` can be slow).

But that's a price that applications like Django, Flask, etc, pay for searching and loading plug-ins at runtime. Hoping it won't slow down Cylc, but I guess we have time to analyze its impact before the final release of Cylc 8 and think in alternatives if necessary?

**Comment by kinow** [comment]
@oliver-sanders the entry point name could be something like `cylc.flow.plugins.main_loop`, which would i) avoid conflicts in case someone ever uses another library that loads entry-points called `main_loop` (unlikely I think, but possible?), and ii) allow us to have something like `cylc.uiserver.plugins.____` too later if we decide to load plug-ins in the UI Server, without causing conflicts.

WDYT?

**Comment by kinow** [comment]
To review, you can either follow the docs or clone this repo https://github.com/kinow/myplugin, then make sure you have the latest version of Cylc 8 `master` which contains the plug-ins. And - using the same virtual/conda environment - install Cylc 8 and then install myplugin with `pip install -e .` (-e optional).

Finally, run a workflow like `cylc run --no-detach five --main-loop=my_plugin` and you should see the hello message in the console, without having to touch Cylc Flow's files or re-install it :crossed_fingers: 

image

**Comment by kinow** [comment]
Added @dwsutherland and @oliver-sanders as reviewers, since we all worked on the previous PR for plug-ins, but in case @hjoliver has time or would like to have a play, feel free to join/replace the reviewers :+1:

**Comment by oliver-sanders** [comment]
Wow I didn’t think it was going to be that simple!

I documented the entry point as ‘cylc.main_loop’ which should avoid conflicts.

Once we’ve sorted out the CLI we should upgrade the scripts entry point to use the same mechanism.

File: cylc/flow/main_loop/__init__.py [hunk 1]
```diff
@@ -311,9 +311,11 @@ class CoroTypes:
 
 def load(config, additional_plugins=None):
     additional_plugins = additional_plugins or []
-    entry_points = pkg_resources.get_entry_map(
-        'cylc-flow'
-    ).get('main_loop', {})
+    entry_points = {
+        entry_point.name: entry_point
+        for entry_point in
+        pkg_resources.iter_entry_points('main_loop')
```
Comment (oliver-sanders) [inline cylc/flow/main_loop/__init__.py:317]:
```suggestion
        pkg_resources.iter_entry_points('cylc.main_loop')
```

Comment (oliver-sanders) [reply cylc/flow/main_loop/__init__.py]: You may also need to update the setup.cfg to match this change.

**Comment by kinow** [comment]
Argh, let me fix that @dwsutherland ! Thanks!

**Comment by kinow** [comment]
- [x] Plugin updated https://github.com/kinow/myplugin/commit/54d8f55f6f4c5579969e4efedb4ef93ac1e0290e
- [x] Entrypoint name updated https://github.com/cylc/cylc-flow/pull/3560/commits/d7e7b16d0ed9976cebccf9cd588ce07676666fb5
- [x] Docs were already using the `cylc.main_loop` @dwsutherland , so didn't need to update that :tada: (https://github.com/cylc/cylc-flow/blob/e0af9f6416a90a8d7d639863b1a118e22bf9635b/cylc/flow/main_loop/__init__.py#L93 & https://github.com/cylc/cylc-flow/blob/e0af9f6416a90a8d7d639863b1a118e22bf9635b/cylc/flow/main_loop/__init__.py#L110)

Ready for review again.

**Pull request merged**